engine/crates/partial-caching/src/execution/defer.rs

```diff
@@ -0,0 +1,174 @@
+            shapes,
+            keys_to_write: vec![],
+            seen_errors: false,
+            output: None,
+        }
+    }
+
+    pub fn query(&self) -> String {
+        self.execution_phase.query()
+    }
+
+    pub fn record_initial_response(&mut self, response: QueryResponse, errors: bool) -> QueryResponse {
+        self.seen_errors = errors;
+
+        let root_shape = self.shapes.root();
+
+        let InitialOutput {
+            mut store,
+            active_defers,
+        } = InitialOutput::new(response, root_shape);
+
+        let mut response_max_age = MaxAge::default();
+
+        if self.execution_phase.has_nocache_partition {
+            // If any portion of our response can't be cached we set the maxAge to none
+            response_max_age.set_none();
+        }
+
+        let cache_entries = self.execution_phase.cache_entries.iter_mut();
+        let cache_keys = std::mem::take(&mut self.execution_phase.cache_keys);
+
+        for (index, (entry, key)) in cache_entries.into_iter().zip(cache_keys).enumerate() {
+            match entry {
+                Entry::Hit(hit, max_age) => {
+                    store.merge_cache_entry(hit, &self.shapes, Some(&active_defers));
+
+                    response_max_age.merge(*max_age);
+                }
+                Entry::Stale(stale) => {
+                    // TODO: Also want to issue an update instruction here, but going to do that
+                    // in GB-6804
+                    store.merge_cache_entry(&mut stale.value, &self.shapes, Some(&active_defers));
+
+                    // This entry was stale so clear the current maxAge until we have revalidated
+                    response_max_age.set_none();
+                }
+                Entry::Miss if key.is_some() => {
+                    response_max_age.merge(Duration::from_secs(
+                        self.execution_phase.cache_partitions[index].0.max_age as u64,
+                    ));
+                    self.keys_to_write.push((key.unwrap(), PartitionIndex(index)));
+                }
+                Entry::Miss => {
+                    response_max_age.merge(Duration::from_secs(
+                        self.execution_phase.cache_partitions[index].0.max_age as u64,
+                    ));
+                }
+            }
+        }
+
+        let return_value = store
+            .reader(&self.shapes)
+            .map(|object| object.into_query_response())
+            .unwrap_or_default();
+
+        self.output = Some(store);
+
+        return_value
+    }
+
+    pub fn record_incremental_response(
+        &mut self,
+        path: &[&QueryPathSegment],
+        data: QueryResponse,
+        errors: bool,
+    ) -> QueryResponse {
+        let Some(output) = &mut self.output else {
+            todo!("GB-6966");
+        };
+
+        if !self.execution_phase.cache_entries.is_empty() {
+            // If we still have cache entries, we should merge them into the
```

Comment:
Into the what?

Reply:
haha, good question.

Reply:
Clearly I am not doing that, oops - will fix